### ISSUE TITLE
Any description of tasmota_counter sensor?

### ISSUE THREAD

**Issue by kajmaj** [hhaim/hass#2]
Hi, your tasmota_sensor looks promising :)
Would it be possible to add any description? 
I've tried to test it, but simple adding of `tasmota_counter.py` seems to be not the proper way.

**Comment by hhaim** [comment]
Have a look into Tasmota wiki/home assistant at the end of the page.
The code is not sync to latest hass version (88.1) so there is a need for a small fix (see the PR for Tasmota switch)
Give me a few days and I will add more info and sync to latest

**Comment by kajmaj** [comment]
Thanks for info. 
I will wait for update, it is a good job anyhow

**Comment by hhaim** [comment]
Checkout README.asciidoc

**Comment by kajmaj** [comment]
So far I use data ( 100 lt = +1) stored in EEPROM. In my case it is 500 writes per year (50 m3 per year/season going though related watermeter) so wemos should last at least a few years.

I´ve tested your counter (watermeter) and it works well but there are a few issues I guess.
- it is necessary to create a rule in tasmota device (Wemos)  to reset counter anyhow as the total accumulated value is limited to 64900 I guess.  
In my case if I consider 4500 pulses per liter it is 140 liters
- it is necessary to set up best suited Counterdebounce, i.e. experiment with the value, to achieve "optimal" average of pulses per liter

Do you work with those two issuaes somehow?

**Comment by hhaim** [comment]
**Good questions**:

There is no need to do **anything** with Tasmota, the custom component should just work. This is the all point. 
It saves the absolute reference in HASS database  and *NOT* in the flash (no need to save anything in flash, no need for any rule).  So even if you have many power losses and wifi disconnections you won't miss even one pulse.

* The Tasmota internal counter is 64bit  (~2E19 pulses) which is a few 1E12 years (even in your case  with 1 cube per day)  -- I believe you will have a power down/reset/watchdog  way before you will get to this number :-)
* The hass code  knows how to handle the case of small Tasmota counter (32bit) but again there is no need for this code 
* The Hass counter is unlimited (python int) 

* Counterdebounce setting -- you should do it **once** and it saved into the flash (I set it to 1000msec) . 
No need for this component to mess with it and it saved over power loss/ OTA update.

Make sense?

**Comment by kajmaj** [comment]
If I understood well, it is all about the math in template.

**Comment by hhaim** [comment]
No. There are two layers of logic that this simple custom component solves.

1. Layer one - manage the absolute total pulses (unlimited) from Tasmota and handle the case of power loss/reboot/wifi discounts etc/counter size/mqtt broker faliure  etc -- no need for flash write in Tasmota -- this is sore in ``hass.tasmota_counter.value`` and it is persistence. 
2. Layer two - convert this absolute number to **litter**. This is why you need the formula in HASS and not in tasmota. In your case, it's something like this  ``140*value/4500 litter``

In short the custom component will do the job with raw Tasmota, there is no need to do anything from your side (just write once the debounce and TelePeroid)
The only thing that can't be done -- is to count pulses while Tasmota has power loss.
 
Let me know if it works for you?

**Comment by hhaim** [comment]
One more thing you can do:
You can checkout AppDaemon app ``CWaterMonitor`` for monitoring the water consumption, notify in cases of irregularity like leakage etc.

**Comment by kajmaj** [comment]
Under the math I meant "second layer" it is easy to set up in Hass.
Counting can be interrupted ( during the measuring cycle) by more reasons:
- lost wemos connection due to home wifi or mqtt failure 
- power lost of at least one participating device, i.e. Wemos, SQL server or entity where HASS runs

I am going to test it more thoroughly.
Will check `CWaterMonitor` as well.
Thank you

**Comment by hhaim** [comment]
This is correct. This component solves the issues you mentioned too. 
This is the nice thing about it - no need for any tricks from Tasmota side. 
I will close this issue and you could open a new one in case you find an issue. 

BTW you could do this check. Go to Tasmota code and change the default value (value after reset) of the counter to be 0xFFFF,FFFF,FFFF,FFFF - 20 
and consume a few liters and check hass value

**Comment by hhaim** [comment]
I’ve looked into Tasmota code. The counter is 32bit and not 64bit. However the conclusions are the same.

**Comment by kajmaj** [comment]
I am testing it at the moment, it is perfect.
Just one question - any possibility to reset your counter sensor?

**Comment by hhaim** [comment]
Why would you want to do that? The Hass counter should match the counter in the water meter. 
a workaround would be to change the formula. 
Another option you can use the "accumulator" custom component that will provide per day/week/month aggregation

**Comment by kajmaj** [comment]
You are right, if I am thinking back - it would not have any sense.

**Comment by hhaim** [comment]
I will try to commit it to hass repo (never did that), let's see how it goes

**Comment by kajmaj** [comment]
Did you succeed with your commitment to the hass repo?
Meanwhile, after update to 88.0 there popped up a warning in hassio:
`Integrations need to be in their own folder. Change sensor/tasmota_counter.py to tasmota_counter/sensor.py. This will stop working soon.`

**Comment by hhaim** [comment]
yes, this is known.
This is part of the change I need to do.  The model has changed now Sensors/Switch/Binary-sensors are under the domain (tasmota)- I think it is better.  So  the new component would be 

```
sensor:
    tasmota:
    type: counter
    ..
```

Will update

**Comment by hhaim** [comment]
Have a look into the new commit -- there was a small change in counter definition

**Comment by kajmaj** [comment]
thanks for update.
Another question:
Accumulator can aggregate different values than on/off time as well?
As you said before - daily, weekly consumption, etc?

**Comment by hhaim** [comment]
It maintained the total time an entity is in a specific state (e.g. the total time a water tap is opened  week/day/month) without running a query in the database (slow)

**Comment by SeByDocKy** [comment]
How to install your tasmota_counter ? For me if I copy the sensor.py into custom_components\tasmota_counter, I got a an error during checking

**Comment by hhaim** [comment]
Just copy all custom_components and copy the usage from pkgs/water

**Comment by SeByDocKy** [comment]
> Just copy all custom_components and copy the usage from pkgs/water

Thanks I managed to install... now checking is ok but I got this error
"the following components and plateforms can not be set up:
-tasmota
-tasmota.switch

Here is my sensor.yaml input

- platform: tasmota
  name: "Eau_chaude_sanitaire"
  stopic: ECS
  id: 1
  max_valid_diff: 2000
  unit_of_measurement: 'l'
  icon: mdi:water-pump
  expire_after: 300
  value_template: "{{ (45497 + (value))|int }}"

My tasmota console infos: 

08:31:50 MQT: tele/ECS/STATE = {"Time":"2019-05-20T08:31:50","Uptime":"0T00:19:08","Vcc":2.786,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":26,"Wifi":{"AP":1,"SSId":"sebydocky","BSSId":"5C:33:8E:CD:27:FC","Channel":6,"RSSI":46,"LinkCount":1,"Downtime":"0T00:00:09"}}
08:31:50 MQT: tele/ECS/SENSOR = {"Time":"2019-05-20T08:31:50","COUNTER":{"C1":10700},"ENERGY":{"TotalStartTime":"2019-05-19T13:44:19","Total":4.751,"Yesterday":2.811,"Today":1.940,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000}}

Got last HASSIO version 0.93.1

Any tips ?

**Comment by hhaim** [comment]
it should be under sensor 
```
sensor:

  - platform: tasmota
    name: Eau_chaude_sanitaire
    stopic: ECS
    id: 1
    unit_of_measurement: 'l'
    icon: mdi:water-pump
    expire_after: 300
    value_template: "{{ (45497 + (value))|int }}"

```

**Comment by SeByDocKy** [comment]
> it should be under sensor
> 
> ```
> sensor:
> 
>   - platform: tasmota
>     name: Eau_chaude_sanitaire
>     stopic: ECS
>     id: 1
>     unit_of_measurement: 'l'
>     icon: mdi:water-pump
>     expire_after: 300
>     value_template: "{{ (45497 + (value))|int }}"
> ```

Yes but was already in sensor.yaml

and of course in configuration.yaml I got
sensor: !include sensor.yaml

**Comment by hhaim** [comment]
In that case it should work. Provide the full error

**Comment by SeByDocKy** [comment]
> In that case it should work. Provide the full error

Where can I found the full long ?  I found nothing in general information of system log

**Comment by hhaim** [comment]
see here
https://community.home-assistant.io/t/var-log-syslog-spammed/10384

**Comment by hhaim** [comment]
My hass version is 89.1. It might be related. let me sync to latest and recheck

**Comment by SeByDocKy** [comment]
Tanks a lot. do i need also to install appdeamon too ? for this particular conpound ?

**Comment by hhaim** [comment]
To use this sensor you just need the custom_component. If you want to leverage the water_app (to find irregularities and notify ) you will need Appdaemon

**Comment by SeByDocKy** [comment]
No I woud like just:

- current last volume of water used and retain this value until the next time faucet will be used
- current debit (l/min) computer from the last usage, retain this value 5s after end of the flow then set to 0
- Total water of the day
- Total water of yesteday....

**Comment by hhaim** [comment]
AppDaemon is not needed

**Comment by kajmaj** [comment]
At the moment I am on  Home Assistant 0.92.2, tasmota does not work

```
Log Details (ERROR)
Mon May 20 2019 17:05:57 GMT+0200 (Středoevropský letní čas)
Unable to prepare setup for platform tasmota: Unable to set up component.

Error during setup of component tasmota
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/homeassistant/setup.py", line 153, in _async_setup_component
    hass, processed_config)
  File "/config/custom_components/tasmota/__init__.py", line 84, in async_setup
    for device in cfg.get(CONF_DEVICES):
TypeError: 'NoneType' object is not iterable
```
Assuming, that error is related to the latest HA changes

**Comment by SeByDocKy** [comment]
Good, the error is reproductable …. I guess should be an easy fix for you

**Comment by hhaim** [comment]
fix is here 
https://github.com/hhaim/hass/commit/44a7b015dc7f7de7810eb7bd2c45828a6348a0aa
I never tested it only with counter sensor ..

**Comment by hhaim** [comment]
BTW it is wasn't related to the hass version (it works with latest) with one sensor now

**Comment by SeByDocKy** [comment]
> fix is here
> 44a7b01
> I never tested it only with counter sensor ..

Thanks :) I will test and report asap I will be home. Actually got one unique water flow sensor…. soon another one will be Added.

**Comment by hhaim** [comment]
After 6 months with the pulse water meter I would buy a RS485 one instead.
There is a need to tune the Debouching value

**Comment by SeByDocKy** [comment]
> RS485

Too many problem with the Tasmota & HASS ?. Actually I also spent some time with the PZEM-004T to monitor my electricity input…. Tasmota serial link can freeze easily. Maybe the pulse counter are already here with ESPhome …. can be a nice alternatve

**Comment by hhaim** [comment]
There are no issues. Pulse is not that accurate because the pulse can debouching.
Tasmota works great. It just an electric property of the pulse. With RS 485(e.g. SDM 630) there is no issue

**Comment by SeByDocKy** [comment]
Ok now it's working no more error..... Now I have to understand how to tune with my count.
I benchmark 400 pulses for 1 liter. How to scale with my own setting ?

I tried: 

value_template: "{{ ((value)|float/400.0)| round(2) }}"

it's not the right value....

**Comment by hhaim** [comment]
Try this
    value_template: "{{ (0 + (value/400.0))|int }}"

**Comment by SeByDocKy** [comment]
Should be more or less ok
I observed that in the scope

value
1419
old value
20551
uptime sec
750

As you can see the "old value" is the total count.... I don't know if your module reset automatically these counting in tasmota ?

**Comment by hhaim** [comment]
Looks good. Tasmota counter value is not persistence. This is the objective of this module

**Comment by SeByDocKy** [comment]
Ok :) .... It's not yet clear in my mind but if it's ok for you. Did you worked a bit how to perform the
- cumsum of today
- cumsum of yesteday ?
and cherry on the cake, how to collect the the debit during an opening/closing cycle ? and the corresponding volume ?

**Comment by hhaim** [comment]
I’m using grafana for that

**Comment by SeByDocKy** [comment]
ho maybe it's possible to share an example with grafana ? I just installed it ... I am also new with ... sorry to bother you :(

**Comment by hhaim** [comment]
Sure, will do

**Comment by hhaim** [comment]
![3810D3A9-CBC1-4888-9E68-99140DF7E6C5](https://user-images.githubusercontent.com/13034429/58070713-c7e36e00-7ba2-11e9-8997-fe15ed665f03.png)
![EACB342C-2DCC-46CF-A546-A924ABC41381](https://user-images.githubusercontent.com/13034429/58070714-c7e36e00-7ba2-11e9-81f7-856660540fba.png)
![C9A3B631-E362-4883-B3F1-5856647DFFC5](https://user-images.githubusercontent.com/13034429/58070716-c87c0480-7ba2-11e9-990c-4cd111e0cb7e.png)

**Comment by hhaim** [comment]
Will write a wiki how to do that

**Comment by SeByDocKy** [comment]
> Will write a wiki how to do that

Whaaoooh look great :)

**Comment by hhaim** [comment]
@SeByDocKy I've started to write it

see  [this](https://github.com/hhaim/hass/wiki/Monitor-water-consumption-and-more-with-Home-Assistant)

please review and edit typos

**Comment by SeByDocKy** [comment]
> @SeByDocKy I've started to write it
> 
> see [this](https://github.com/hhaim/hass/wiki/Monitor-water-consumption-and-more-with-Home-Assistant)
> 
> please review and edit typos


Ho it's great :) .... Yesteday I already started to do my first grafana plots .... tried to played with derivative as well to collect instantaneous usage.... but your precedure seems more polished. 
I also use a mix with your custom components and force my Counter1 to reset at 00:00::00 everyday.... via automatisation. I also store into a input_number the daily usage..... what's the hell with HASS, there is no way to have the method set_value with the template sensor class..... Need to trick that with input sensor. Generally spearking, HASS should integrate some "virtual template sensor" who can be updated also via automatisation....

**Comment by hhaim** [comment]
It is not a good idea to **reset** the counter
The component handles the counter and you should **not** touch it. 
Influxdb has a lot of capabilities (e.g. integrals and diffs) so with **one** counter you could have any type of chart (per day, per week, per month etc)

**Comment by SeByDocKy** [comment]
> It is not a good idea to **reset** the counter
> The component handles the counter and you should **not** touch it.
> Influxdb has a lot of capabilities (e.g. integrals and diffs) so with **one** counter you could have any type of chart (per day, per week, per month etc)

Why it's not a good idea to reset the counter ? and generally what is used for the debounce value in your tasmota settings ? to filter/integrate pulses versus time ? to smooth results ?

I had to create the yesteday variable in HASS in order to plot the cumsum of water in a water with the yesteday constant value in a second horizontal line....

What I woud like is to is to count per hour or day, the number of time the tape is opened then closed (cyle open=> close).... so mathematically something like   grad( abs(grad(value) > threshold ) > 0.5 versus time ..... then just integrate that over 1h, 1 day... etc.... it should count the number of opening-closing cylcles

**Comment by hhaim** [comment]
There is no need to do that, copy the query I did and see how it works

**Comment by SeByDocKy** [comment]
The images associated with the query of the water burst is not displayed correctly in your wiki :)

**Comment by hhaim** [comment]
fixed

**Comment by SeByDocKy** [comment]
and for the Water flow speed query, it shouldn't be the water_total_norm as entiy_id ? (expressed in liter) here it's water_total (pulses)

**Comment by hhaim** [comment]
yep, you are right ..
You can edit the wiki and add this notes

**Comment by SeByDocKy** [comment]
Now I am in the installation process of all new elements
In copied in my configuration.yaml

water_monitor:
  module: heat_app
  class: CWaterMonitor
  sensor_water_total: sensor.water_total # input: sensor of water in pulses (not litter)
  sensor_water_leak_detector: variable.water_leak_detector # output: filter one pulse
  sensor_water_bursts: variable.water_bursts # output: filter burst in litters
  watchdog_duration_min: 50 # input: high watermark  in minutes to open a tap
  watchdog_leakage_ticks: 3 # input: how many pulses is a leakage?
  max_day: 2000 # input: max consumption a day
  max_burst_l0: 400  # input: max burst in litters
  max_burst_l1: 600  # input: max burst in litters  when garden taps are open
  taps_switchs: [ switch.wbi_p1, switch.wbi_p2] # when one of them is enabled move to burstl1

but got an error 
Configuration invalid
Integration not found: water_monitor

I didn't see any water_monitor folder in custom_conponents ....

**Comment by hhaim** [comment]
This is AppDaemon part not HA part.  search in my project for the location of it

**Comment by SeByDocKy** [comment]
> This is AppDaemon part not HA part. search in my project for the location of it

 
Was already installed for me .... but still can't see this integration. I also copied heat_app.py in appdaemon\apps .... didn't help too :(

**Comment by hhaim** [comment]
What is the error?

**Comment by hhaim** [comment]
Please open a different ticket
